automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala

```diff
@@ -23,110 +23,115 @@ class BillingSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Cl
+         *
+         * A was a subset of B, C and D
+         * C was a subset of D
+         *
+         * Clearly a lot of functionality bleeds between these test cases. In order to save runtime during
+         * the tests, we're combining all five. The idea is that if any piece of basic functionality
+         * in these tests was broken (creating a workspace in a new project, creating a new project, etc), then
+         * most of these tests would fail anyway so grouping them together isn't much worse.
+         *
+         * NOTE: this single test case MUST create a brand new billing project each time it's run. If we use a
+         * project from GPAlloc, this entire test becomes worthless and our test coverage for billing is garbage.
+         *
+         * E comes last because it makes the project unusable.
+         */
+        "should be able add a new user, create a workspace, and run a method, change billing account" in {
           val user = UserPool.chooseProjectOwner
           implicit val authToken: AuthToken = user.makeAuthToken()
+          val secondUser = UserPool.chooseStudent.email
 
-          withBillingProject("billing-spec-make-ws") { billingProjectName =>
-            withWebDriver { implicit driver =>
-              withSignIn(user) { listPage =>
-                // create workspace and verify
-                val workspaceName = "BillingSpec_makeWorkspace_" + randomUuid
-                register cleanUp api.workspaces.delete(billingProjectName, workspaceName)
-                val detailPage = listPage.createWorkspace(billingProjectName, workspaceName)
-              }
-            }
-          }
-        }
-
-        "should be able to run a method in a new workspace in the billing project" in {
-          val user = UserPool.chooseProjectOwner
-          implicit val authToken: AuthToken = user.makeAuthToken()
 
           /*
            * This must continue to create a new billing project rather than using an allocated one.
            * Otherwise it's not covering the entirety of the intended scenario (which is one that we
            * have seen break in the past that relied on using a brand new billing project).
            */
-          withBillingProject("billing-spec-method") { billingProjectName =>
-            withWorkspace(billingProjectName, "BillingSpec_runMethod") { workspaceName =>
+          withWebDriver { implicit driver =>
+            withSignIn(user) { listPage =>
+              //BEGIN: Test creating billing project in UI
+              val billingPage = new BillingManagementPage().open
+              val billingProjectName = createNewBillingProject(user, billingPage)
+              //END: Test creating billing project in UI
+
+              //BEGIN: Test creating workspace
+              val workspaceName = "BillingSpec_makeWorkspace_" + randomUuid
+              register cleanUp api.workspaces.delete(billingProjectName, workspaceName)
```

Comment:
All new to me, and I don't have the source for the `CleanUp` trait yet.  However, I'm going to assume that this trait knows how to respond to just about any failure anywhere in this test to ensure that any and all created assets are properly cleaned up after a successful run or any failure?  

Without knowing details, is there anything else we can or should be cleaning up out of this test?  Users?  Buckets?

Reply:
look in the workbench-libs repo for the source.

cleaning up the workspace cleans up the bucket (and so does cleaning up the project which is below). Users are cleaned when the fiab is torn down (fiab-stop).